src/irmin-pack/IO.ml

```diff
@@ -232,25 +259,27 @@ module Unix : S = struct
     mkdir (Filename.dirname file);
     match Sys.file_exists file with
     | false ->
-        let x = Unix.openfile file Unix.[ O_CREAT; mode; O_CLOEXEC ] 0o644 in
-        let raw = Raw.v x in
-        Raw.unsafe_set_offset raw 0L;
-        Raw.unsafe_set_version raw current_version;
-        v ~offset:0L ~version:current_version raw
+        if readonly then v ~offset:0L ~version:current_version None
+        else
+          let x = Unix.openfile file Unix.[ O_CREAT; mode; O_CLOEXEC ] 0o644 in
+          let raw = Raw.v x in
+          Raw.unsafe_set_offset raw 0L;
+          Raw.unsafe_set_version raw current_version;
+          v ~offset:0L ~version:current_version (Some raw)
     | true ->
         let x = Unix.openfile file Unix.[ O_EXCL; mode; O_CLOEXEC ] 0o644 in
         let raw = Raw.v x in
         if fresh then (
           Raw.unsafe_set_offset raw 0L;
           Raw.unsafe_set_version raw current_version;
-          v ~offset:0L ~version:current_version raw )
+          v ~offset:0L ~version:current_version (Some raw) )
         else
-          let offset = Raw.unsafe_get_offset raw in
+          let offset = if readonly then 0L else Raw.unsafe_get_offset raw in
```

Comment:
This change isn't clear to me; why the different behaviour for RO instances if we know that `raw` exists? (is it just an unrelated optimisation?)

Reply:
It's related to https://github.com/mirage/irmin/pull/955/files#diff-1beb7e2142825f197755316a39d00699R227 : before, when an ro opened an existing file, it refilled its buffers with what was in that file starting from offset `0`. 
Now, an ro cannot do a refill when opening a file (as it might not exists) and so it needs to do it when it first reads a value. At that point, it compares its offset with the real offset (https://github.com/mirage/irmin/pull/955/files#diff-1beb7e2142825f197755316a39d00699L165). To ensure that it does sync with the on-disk file, we have to set the offset to `0` when opening the file in ro. Does it make sense?  

Agree that it is not clear, I should probably add a comment in the code about this.

Reply:
I think I understand, thanks :) However, I think in this particular case we _know_ that the file exists (since we've just opened it), right?

Reply:
Yes, but if you want to check that the file exists from `irmin-pack.ml` you need to add a `if Sys.file_exists` in the code here https://github.com/mirage/irmin/pull/955/files#diff-1beb7e2142825f197755316a39d00699R227 (which I don't want to do, I think it breaks a bit the abstraction of file). 
By having the offset set to `0` for all ro instances when they are created, we avoid checking this. We have to do `refill` anyway, it is just postponed until the first read.

Reply:
I'm not proposing changing it in `irmin-pack.ml`, just here where it's not strictly necessary.

As you point out, it doesn't really make much of a difference anyway. Let's not worry about it :slightly_smiling_face: